solve() and solveset(): equations with non-commutative symbols

valglad opened the pull request:
Currently solve() and solveset() return incorrect solutions for
linear equations with non-commutative symbols. This fixes it.

Fixes #12258 

<!-- BEGIN RELEASE NOTES -->
* solvers
    * implemented `nc_solve`, a simple solver for non-commutative equations; non-commutative equations no longer produce incorrect solutions
<!-- END RELEASE NOTES -->

valglad committed:
solve() and solveset(): equations with non-commutative symbols

Currently solve() and solveset() return incorrect solutions for
linear equations with non-commutative symbols. This fixes it.

valglad committed:
corrected typo in test_solveset.py

asmeurer commented:
But does this still give the right answer for `solve(x*M+N*y, x)`?

asmeurer commented:
I think there really needs to be completely separate solvers for noncommutatives. The solve code will assume commutative expressions in too many places. It would be better to have a completely distinct noncommutative solver algorithm that always does the right thing with noncommutative expressions, even if it is more limited in power.

valglad commented:
Hmm, that is true. But wouldn't writing a non-commutative solve amount to essentially copying the solve code that's there already and carefully changing all of the commutative assumptions anyway?
That would probably take a lot of time. I could try writing a non-commutative solver for linear equations like above and throw a NotImplemented error otherwise for now. Would that be useful?

smichr commented:
> But does this still give the right answer for solve(x*M+N*y, x)?

I'm guessing you mean something like `solve(X*M + N*y, X)` where `X` is non-commutative. When solving for a commutative symbol it won't matter on which "side" the noncommutative `M` is located.

smichr commented:
> needs to be completely separate solvers for noncommutatives

In the example here, however, one is solving for a commutative symbol while moving the non-commutative part by inversion. So I think what is being done is ok. When solving for a non-commutative symbol one would have to handle pre and post-multiplies differently, e.g. `solve(L*X*R - rhs, X) -> 1/L*rhs*1/R`. `.args_cnc()` should help with the process:

```
>>> var('A:C', commutative=False)
(A, B, C)
>>> (A*B*C*x).args_cnc()
[[x], [A, B, C]]
```

smichr commented:
I'm with @asmeurer on the idea of being nervous about this working for all expressions. I would need some time or evidence to indicate that you've found the only places where division of factors away from the symbol of interest might involve noncommutative symbols. 

I see that polys raises an error for such expressions:

```
>>> solve(x**2+N*x-3,x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\solvers\solvers.py", line 1053, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "sympy\solvers\solvers.py", line 1408, in _solve
    poly = Poly(f_num)
  File "sympy\polys\polytools.py", line 90, in __new__
    return cls._from_expr(rep, opt)
  File "sympy\polys\polytools.py", line 200, in _from_expr
    rep, opt = _dict_from_expr(rep, opt)
  File "sympy\polys\polyutils.py", line 352, in _dict_from_expr
    raise PolynomialError('non-commutative expressions are not supported')
sympy.polys.polyerrors.PolynomialError: non-commutative expressions are not supported
```

valglad commented:
Yes, I agree what I initially did isn't enough. I've written a piece of code to handle linear non-commutative equations in one variable and am currently trying to integrate it into the main solve. So I think I'm going to check if there are any non-commutative symbols in a given equation using `args_cnc` and if there are, let the function that I've written deal with it if it can, otherwise throw `NotImpementedError` (like for the example that you gave above, because it's not linear).

valglad committed:
a solver function for equations with noncommutative symbol

This commit introduces a new function `nc_solve()` designed to deal
with linear equations in one variables involving noncommutative symbols.
A NotImplementedError is returned for other types of equations and
inequalities.

valglad committed:
Merge branch 'master' into nc_solve

valglad committed:
Merge branch 'master' into nc_solve

valglad committed:
Merging `master` properly

valglad committed:
another attempt at merging `master`

valglad committed:
Merge branch 'nc_solve' of github.com:valglad/sympy into nc_solve

valglad commented:
Sorry there are so many merging commits, it took me a while to get it right.

smichr commented on sympy/solvers/solvers.py:
```diff
@@ -1846,6 +1862,57 @@ def _ok_syms(e, sort=False):
+    if not (expr.is_Add):
+        left, expr = expr.as_independent(symbol, as_Add=False)
+        right = 1/symbol*expr
+        return [-1/left*ind*1/right]
```
Won't `ind` just be 0 and hence the result 0 in this case?

valglad replied:
No, because the `if` block preceding it changes what `expr` is. For example, consider solving `expr = M*x+a` in `x`. It enters `if` on line 1890 since `expr.is_Add = True` which gives `ind = a` and `expr = M*x`. Now `expr.is_Add = False` so it enters the next if and the returned result is `-M**(-1)*a`, not 0.

valglad commented:
@smichr  I get the error in test_integrals even in my master branch and `git pull` says it's up-to-date. However, my expr.py file seems to be different from sympy/master. I'm confused. Isn't `git pull` supposed to take care of that?

valglad committed:
updated expr.py

asmeurer commented:
For the original equation from the issue, everything is noncommutative (A, B are matrices and x, y are vectors). https://github.com/sympy/sympy/issues/7379 is related.

asmeurer commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1823,7 +1823,13 @@ def test_issue_12114():
+
+def test_issue_12258():
+    M, N, x, y = symbols('M N x y',commutative=False)
+    assert solve(M*x+N*y,x) == [-M**(-1)*N*y]
```
Add tests for both sides.

asmeurer commented on sympy/solvers/solvers.py:
```diff
@@ -1849,6 +1865,57 @@ def _ok_syms(e, sort=False):
+        if all(map(lambda x: x == rights[0],rights)):
+            left = sum(lefts)
+            right = rights[0]
+            return [-1/left*ind*1/right]
```
I would avoid using `1/left`, preferring `left**-1` for noncommutatives.

asmeurer commented:
The way `solveset` is designed, maybe it would make more sense to have an `ncsolveset`. One issue is that you cannot so easily say that you have found all the solutions for noncommutative equations. Even A*B = 0 -> A = 0 or B = 0 doesn't hold for noncommutative (the reverse still holds, but you can have zero divisors, e.g., two nonzero matrices that multiply to zero).

But let's open a new issue to discuss this once this is merged.

asmeurer commented:
CC @sympy/sympy-solvers

valglad committed:
added more tests and replaced 1/left with left**(-1)

valglad commented:
> Even A*B = 0 -> A = 0 or B = 0 doesn't hold for noncommutative

It does if we are going to assume that everything non-zero has a multiplicative inverse which the description of the problem was kind of assuming. Because if we are just given `M*x + N*y` and no information about whether or not M is invertible, we can't just return `-M**(-1)*N*y ` as the issue post suggests. If all symbols had the attribute `is_invertible`, this could be made more general.

asmeurer commented:
Anyway, this does fix a wrong answer. So we can discuss the separate ncsolve in a new issue.

@smichr, @sympy/sympy-solvers do you have any objections here?

valglad commented:
@smichr @asmeurer This is an almost complete PR that's been around for a while

Shekharrajak commented:
I can see the conflicts. Please fix the conflicts first @valglad .

valglad committed:
Merge branch 'master' into nc_solve

czgdp1807 commented:
ping @smichr @asmeurer Please provide your views on this PR. The changes LGTM.

czgdp1807 commented:
Thanks @valglad for the contributions. It looks like the associated issue has been fixed. Refer, https://github.com/sympy/sympy/issues/12258#issuecomment-557181780
Closing this for now. Please let me know if it needs to be opened again.
